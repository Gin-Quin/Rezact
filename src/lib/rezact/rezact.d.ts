declare var xCreateElement: any;
declare var xFragment: any;
