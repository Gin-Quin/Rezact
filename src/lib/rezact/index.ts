let createElement, createTextNode, createComment, createDocumentFragment;
if (typeof window === "object") {
  createElement = document.createElement.bind(document);
  createTextNode = document.createTextNode.bind(document);
  createComment = document.createComment.bind(document);
  createDocumentFragment = document.createDocumentFragment.bind(document);
}
export { createElement, createTextNode, createComment, createDocumentFragment };
export const isArray = Array.isArray;

export let createComponent = (tagName, attributes = null) =>
  tagName(attributes);

export const setCreateCompFunc = (func) => (createComponent = func);

export function xCreateElement(tagName, attributes, ...children) {
  if (tagName === xFragment) {
    (children as any).rezactFragment = true;
    return children;
  }
  if (typeof tagName === "function") {
    attributes = attributes || {};
    attributes.children = attributes.children || children;

    let componentContext = null;
    attributes.setContext = (key, context) => {
      if (!componentContext) componentContext = {};
      componentContext[key] = context;
    };

    attributes.getContext = (key) => {
      let currentElement = contextRoot;

      while (currentElement.parentNode) {
        currentElement = currentElement.parentNode;

        if (
          currentElement.firstChild &&
          currentElement.firstChild.rezactContext &&
          currentElement.firstChild.rezactContext[key]
        ) {
          return currentElement.firstChild.rezactContext[key];
        }

        if (currentElement.rezactContext && currentElement.rezactContext[key]) {
          return currentElement.rezactContext[key];
        }
      }

      return undefined;
    };

    const newComp = createComponent(tagName, attributes);

    const contextRoot = newComp[0] || newComp;
    contextRoot.rezactContext = componentContext;

    return newComp;
  }
  const elm = createElement(tagName);
  if (attributes) handleAttributes(elm, attributes);
  const childLen = children.length;
  for (let i = 0; i < childLen; i++) {
    appendChild(elm, children[i]);
  }
  return elm;
}

export let attributeHandlers = [];

export const addAttributeHandler = (item) => attributeHandlers.unshift(item);

const evKeys = {
  onClick: "click",
  onDblClick: "dblclick",
  onKeyDown: "keydown",
  onSubmit: "submit",
  onChange: "change",
  onInput: "input",
};
function handleAttributes(elm, attrs) {
  const keys = Object.keys(attrs);
  const keyLen = keys.length;
  outer: for (let i = 0; i < keyLen; i++) {
    const key = keys[i];
    const attrVal = attrs[key];
    if (evKeys[key] && typeof attrs[key] === "function") {
      elm.addEventListener(evKeys[key], attrVal);
      continue;
    }
    if (typeof attrVal === "boolean" && attrVal) {
      elm.setAttribute(key, "");
      continue;
    }
    const attrHandlerLen = attributeHandlers.length;
    for (let x = 0; x < attrHandlerLen; x++) {
      const hook = attributeHandlers[x];
      if ((hook.matches as any)(attrs, key, attrVal)) {
        (hook.handler as any)(elm, key, attrVal, attrs);
        continue outer;
      }
    }

    elm.setAttribute(key, attrVal);
  }
}

export const childArrayHandler: any = {
  matches: (child) => isArray(child),
  handler: (parent, child) => {
    const childLen = child.length;
    for (let i = 0; i < childLen; i++) {
      appendChild(parent, child[i]);
    }
  },
};

export const childNodeHandler: any = {
  matches: (child) => child instanceof Node,
  handler: (parent, child) => parent.appendChild(child),
};

let appendChildHooks: any = [childArrayHandler, childNodeHandler];

export const addAppendChildHook = (item) => appendChildHooks.unshift(item);

export let appendChild = (parent, child, ...args) => {
  const appendChildHookLen = appendChildHooks.length;
  if (
    typeof child === "undefined" ||
    child === null ||
    typeof child === "boolean"
  )
    return;

  for (let i = 0; i < appendChildHookLen; i++) {
    const hook = appendChildHooks[i];
    if ((hook.matches as any)(child))
      return (hook.handler as any)(parent, child, ...args);
  }

  parent.appendChild(createTextNode(String(child)));
};

const afterRenderHooks = [];
export const addAfterRenderHook = (item) => afterRenderHooks.push(item);

export function render(root, tagName, attributes: any = {}) {
  const elm = createComponent(tagName, attributes);
  appendChild(root, elm);
  afterRenderHooks.forEach((func) => func());
}
export const xFragment = [];

export let handleInputValue = null;
export function useInputs() {
  if (handleInputValue) return;
  handleInputValue = true;
  function getInputVal(elm: HTMLInputElement) {
    const radioVal = elm.value || elm.id;
    if (elm.type === "radio" && elm.checked) return radioVal;
    if (elm.type === "radio" && !elm.checked) return "";
    if (elm.type === "checkbox") return elm.checked;
    if (elm.type === "number") return +elm.value;
    if (elm.value) return elm.value;
    return "";
  }

  function setInputVal(elm: any, val: any) {
    const radioVal = elm.value || elm.id;
    if (elm.type === "radio" && val === radioVal) return (elm.checked = true);
    if (elm.type === "radio" && val !== radioVal) return (elm.checked = false);
    if (elm.type === "checkbox") return (elm.checked = !!val);
    elm.value = val.textContent || val;
  }

  const handleInputAttr = (element, attributeValue, attributes) => {
    setInputVal(element, attributeValue.get());

    attributeValue.subscribe(
      (newVal: string) => {
        if (element.value === newVal) return;
        setInputVal(element, newVal);
      },
      { elm: element }
    );

    if (
      !Object.keys(attributes).includes("onChange") &&
      !Object.keys(attributes).includes("onInput")
    ) {
      const inpEvType = element.type === "text" || element.type === "number";
      const evType = inpEvType ? "input" : "change";
      element.addEventListener(evType, () => {
        attributeValue.set(getInputVal(element));
      });
    }
  };

  const attributeInputValueHandler = {
    matches: (_attrs, key, attrVal) =>
      (key === "value" && attrVal.state) ||
      (key === "checked" && attrVal.state),
    handler: (elm, _key, attrVal, attrs) =>
      handleInputAttr(elm, attrVal, attrs),
  };

  addAttributeHandler(attributeInputValueHandler);
}

let usingLifeCycleAttrs = false;
export function useLifeCycleAttributes() {
  if (usingLifeCycleAttrs) return;
  usingLifeCycleAttrs = true;

  const onMountHandler = (elm, _key, attrVal, _attrs) => {
    const checkMounted = () => {
      if (elm.isConnected) return (elm.rzHasMounted = true) && attrVal(elm);
      setTimeout(checkMounted, 10);
    };
    checkMounted();
  };

  const onMountAttrHandler = {
    matches: (_attrs, key) => key === "onMount",
    handler: onMountHandler,
  };

  const onUnMountAttrHandler = {
    matches: (_attrs, key) => key === "onUnmount",
    handler: (elm, _key, attrVal, attrs) => {
      if (!attrs.onMount) onMountHandler(elm, _key, () => {}, attrs);

      const checkMounted = () => {
        if (elm.rzHasMounted && !elm.isConnected) return attrVal(elm);
        setTimeout(checkMounted, 100);
      };
      checkMounted();
    },
  };

  addAttributeHandler(onMountAttrHandler);
  addAttributeHandler(onUnMountAttrHandler);
}
