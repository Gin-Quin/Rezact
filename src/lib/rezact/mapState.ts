import { BaseState, createComputed } from "./signals";
import {
  addAppendChildHook,
  appendChild,
  childArrayHandler,
  childNodeHandler,
  createComment,
  createComponent,
  createDocumentFragment,
  createElement,
  isArray,
} from ".";

export class MapState extends BaseState {
  constructor(st: any) {
    super(st);
    ["push", "pop", "splice", "shift", "unshift"].forEach((item) => {
      this[item] = (...args) => {
        this.value[item](...args);
        this.updateList(this.func);
        this.alertSubs(this.value);
      };
    });
  }

  Map(func: any) {
    if (!func.args) {
      this.func = func;
      const _args = func.toString().split("=>")[0].replaceAll(" ", "");
      const args = _args.slice(1, _args.length - 1).split(",");
      func.args = args;
    }
    this.updateList(func);
    return this.value;
  }

  updateList(func) {
    const len = this.value.length;
    for (let idx = 0; idx < len; idx++) {
      let item = this.value[idx];
      if (!item.state) this.value[idx] = item = new BaseState(item);
      item.__private_idx = idx;
      if (func.args[1]) {
        if (!item.idxState) {
          item.idxState = new BaseState(0);
        }
        item.idxState.setValue(idx);
      }

      if (item.elmRef?.isConnected === false) item.elmRef = undefined;
      item.elmRef = item.elmRef || createComponent({ func, item });

      isArray(item.elmRef)
        ? item.elmRef.forEach((elm) => (elm.associatedState = item))
        : (item.elmRef.associatedState = item);
    }
  }

  deleteValue(valToDelete: any) {
    const index = this.value.indexOf(valToDelete);
    if (index < 0) return;
    if (valToDelete.elmRef) valToDelete.elmRef.remove();
    this.value.splice(index, 1);
    setTimeout(() => this.updateList(this.func), 0);
  }

  toJson() {
    return this.value.map((thisVal) =>
      thisVal.value !== undefined && thisVal.value !== null
        ? thisVal.getValue()
        : thisVal
    );
  }
}

const addChildren = (values: any, parentNode) => {
  if (values === undefined) return;
  const len = values.length;
  let nextNode = parentNode;
  for (let i = 0; i < len; i++) {
    const elm = values[i].elmRef;
    if (!(nextNode.nextSibling === (elm[0] || elm))) {
      appendChild(nextNode, elm, true);
    }
    nextNode = isArray(elm) ? elm.at(-1) : elm;
  }
};

const remPlaceHolder = createElement("span");
const removeStaleChildren = (parentNode, endNode, parent, child) => {
  const values = child.value;
  let nextNode = parentNode.nextSibling;

  if (
    parentNode.parentNode.childNodes.length === child.previousChildLen + 2 &&
    values.length === 0
  ) {
    return (parentNode.parentNode.innerHTML = "");
  }

  if (values.length === 0) {
    let frag = createDocumentFragment();
    const placeholder = createElement("span");
    parent.parentNode.insertBefore(placeholder, parent);
    frag.appendChild(parent);
    while (nextNode !== endNode) {
      const thisNode = nextNode;
      nextNode = nextNode.nextSibling;
      thisNode.remove();
    }
    placeholder.parentNode.insertBefore(frag, placeholder);
    parent.appendChild(frag);
    placeholder.remove();
    return;
  }

  let elmRefIdx = 0;
  const childLen = values[0].elmRef.length || 1;
  let inc = 0;
  while (nextNode !== endNode) {
    const idx = Math.floor(inc / childLen);
    elmRefIdx = inc % childLen;
    const val = values[idx];
    let elmRef = val && val.elmRef && (val?.elmRef[elmRefIdx] || val?.elmRef);

    if (values.indexOf(nextNode.associatedState) < 0) {
      nextNode = nextNode.nextSibling;
      const prevElm = nextNode.previousSibling;
      if (prevElm.replaceWith && elmRef instanceof HTMLElement) {
        prevElm.replaceWith(elmRef);
        nextNode = elmRef.nextSibling;
      } else {
        prevElm.remove();
      }
    } else if (val && elmRef && nextNode !== elmRef) {
      nextNode = nextNode.nextSibling;
      const prevElm = nextNode.previousSibling;
      if (
        prevElm.replaceWith &&
        elmRef instanceof HTMLElement &&
        elmRef.parentNode
      ) {
        elmRef.replaceWith(remPlaceHolder);
        prevElm.replaceWith(elmRef);
        remPlaceHolder.replaceWith(prevElm);
      } else {
        elmRef.remove();
      }
    } else {
      nextNode = nextNode.nextSibling;
    }
    inc += 1;
  }
};

const handleArray = (parent: any, child: any) => {
  const frag = createDocumentFragment();
  const frac = frag.appendChild.bind(frag);
  const parentNode = createComment("start map");
  const endNode = createComment("end map");

  frac(parentNode);
  frac(endNode);

  child.subscribe((newVal: any) => {
    if (child.previousChildLen === 0) {
      frac(parentNode);
      frac(endNode);
    }
    removeStaleChildren(parentNode, endNode, parent, child);
    addChildren(newVal, parentNode);
    if (child.previousChildLen === 0) parent.appendChild(frag);
    child.previousChildLen = child.value.length;
  });

  child.previousChildLen = child.value.length;
  addChildren(child.value, parentNode);
  parent.appendChild(frag);
};

const childStateHandler = {
  matches: (child) => isArray(child.value) && child.state,
  handler: (parent, child) => handleArray(parent, child),
};

addAppendChildHook(childStateHandler);

function insertNodeAfter(currentNode: any, childNode: any) {
  if (currentNode.nextSibling) {
    currentNode.parentNode.insertBefore(childNode, currentNode.nextSibling);
  } else {
    currentNode.parentNode.appendChild(childNode);
  }
}

function appendChildNode(
  parentNode: any,
  childNode: any,
  insertAfter: boolean = false,
  removeElm: boolean = false
) {
  if (removeElm) return childNode.remove();
  insertAfter
    ? insertNodeAfter(parentNode, childNode)
    : parentNode.appendChild(childNode);
}

childNodeHandler.handler = (parent, child, insertAfter, removeElm) =>
  appendChildNode(parent, child, insertAfter, removeElm);

childArrayHandler.handler = (parent, child, insertAfter, removeElm) => {
  const len = child.length;
  if (insertAfter && isArray(child)) {
    for (let i = len - 1; i > -1; i--) {
      appendChild(parent, child[i], insertAfter, removeElm);
    }
  } else {
    for (let i = 0; i < len; i++) {
      appendChild(parent, child[i], insertAfter, removeElm);
    }
  }
};

export const createMapped = (func, deps) => {
  const computed: any = createComputed(func, deps);
  computed.deps = deps;
  return computed;
};
