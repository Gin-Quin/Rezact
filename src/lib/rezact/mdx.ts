import { xCreateElement, xFragment } from ".";

export default { xCreateElement, xFragment };
