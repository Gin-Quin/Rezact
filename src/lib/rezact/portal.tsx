import {
  // addPostCreateComponentHook,
  // addPreCreateComponentHook,
  render,
} from ".";

// addPreCreateComponentHook((tagName, attributes, hookContext) => {
//   console.log({ tagName, attributes, hookContext });
// });
// addPostCreateComponentHook((newComp, tagName, attributes, hookContext) => {
//   console.log({ newComp, tagName, attributes, hookContext });
// });

export function createPortal(elm) {
  return function Portal(props: any) {
    const { children } = props;
    const portalMount = () => {
      if (typeof elm === "string") elm = document.querySelector(elm);
      elm.dataset.portal = "out";
      elm.innerHTML = "";

      render(elm, () => children);
      portalIn.rzPortalOut = elm;
      elm.rzPortalIn = portalIn;
    };

    const portalIn = <span data-portal="in" onMount={portalMount} />;
    return portalIn;
  };
}
