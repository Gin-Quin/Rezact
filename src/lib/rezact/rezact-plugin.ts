import { PluginOption } from "vite";
import { Parser } from "acorn";
import jsx from "acorn-jsx";
import * as walk from "acorn-walk";
import MagicString from "magic-string";

const acorn = Parser.extend(jsx());

let src = "";
let magicString: any = null;
let lastImport: any = null;
let importsUsed: any = {};
let signalsUsed: any = {};
let mapStateUsed: any = {};
let functionsToRun: any = [];

function wrapInUseSignal(node) {
  signalsUsed.BaseState = true;
  magicString.appendLeft(node.start, `new BaseState(`);
  magicString.appendRight(node.end, `)`);
}

function wrapInUseMapState(node) {
  mapStateUsed.MapState = true;
  magicString.appendLeft(node.start, `new MapState(`);
  magicString.appendRight(node.end, `)`);
}

function findDependencies(node, excludeDeps = {}) {
  const assignments: any = { ...excludeDeps };
  const identifiers: any = {};
  walk.simple(node, {
    AssignmentExpression(node: any) {
      if (node.left.name && node.left.name[0] !== "$") return;
      assignments[node.left.name] = true;
    },

    UpdateExpression(node: any) {
      if (node.argument.name && node.argument.name[0] !== "$") return;
      assignments[node.argument.name] = true;
    },

    Identifier(node: any) {
      if (node.name && node.name[0] !== "$") return;
      identifiers[node.name] = true;
    },
  });
  const ids = Object.keys(identifiers);
  const assigns = Object.keys(assignments);
  return ids.filter((id) => !assigns.includes(id));
}

function wrapInCreateComputed(node, explicitDeps = null, excludeDeps = {}) {
  const deps = explicitDeps || findDependencies(node, excludeDeps);
  if (deps.length === 0) return;
  signalsUsed.createComputed = true;
  magicString.appendLeft(
    node.start,
    `createComputed(([${deps.join(",")}]) => `
  );
  magicString.appendRight(node.end, `, [${deps.join(",")}])`);
}

function wrapInCreateComputedAttribute(
  node,
  explicitDeps = null,
  excludeDeps = {}
) {
  const deps = explicitDeps || findDependencies(node, excludeDeps);
  if (deps.length === 0) return;
  signalsUsed.createComputedAttribute = true;
  magicString.appendLeft(
    node.start,
    `createComputedAttribute(([${deps.join(",")}]) => `
  );
  magicString.appendRight(node.end, `, [${deps.join(",")}])`);
}

function wrapInCreateMapped(node, explicitDeps = null, excludeDeps = {}) {
  mapStateUsed.createMapped = true;
  const deps = explicitDeps || findDependencies(node, excludeDeps);
  if (deps.length === 0) return;
  signalsUsed.createComputed = true;
  magicString.appendLeft(node.start, `createMapped(([${deps.join(",")}]) => `);
  magicString.appendRight(node.end, `, [${deps.join(",")}])`);
}

function tackOnDotVee(node) {
  magicString.appendRight(node.end, `.getValue()`);
}

function hasAncestor(ancestors, type) {
  return ancestors.map((a) => a.type).indexOf(type) > -1;
}

function inFunction(ancestors, funcName) {
  for (let i = ancestors.length - 1; i > -1; i--) {
    const anc = ancestors[i];
    if (anc.type === "CallExpression" && anc.callee.name === funcName)
      return anc;
  }
  return false;
}

function isAttributeArg(ancestors) {
  const func = inFunction(ancestors, "xCreateElement");
  if (!func) return false;
  let arg = null;
  for (let i = ancestors.length - 1; i > -1; i--) {
    const anc = ancestors[i];
    if (anc.type === "ObjectExpression") {
      arg = anc;
      break;
    }
  }
  if (func.arguments[1] === arg) return true;
  return false;
}

function wrapInSetValue(node, nestedMember = false) {
  if (nestedMember) {
    const left = src.slice(node.left.start, node.left.end);
    const right = src.slice(node.right.start, node.right.end);

    const backupMagicString = magicString;
    const backupMagicSrc = src;
    const leftAst = acorn.parse(left, {
      locations: true,
      ecmaVersion: "latest",
      sourceType: "module",
    });
    src = left;
    magicString = new MagicString(left);
    compileRezact(leftAst);
    const leftParsed = magicString.toString();
    const leftMinusValue =
      leftParsed.slice(-11) === ".getValue()"
        ? leftParsed.slice(0, leftParsed.length - 11)
        : leftParsed;

    magicString = backupMagicString;
    src = backupMagicSrc;

    let nestedRightVal = right || "";
    if (node.operator === "+=") nestedRightVal = `${leftParsed} + ${right}`;
    magicString.overwrite(
      node.start,
      node.end,
      `${leftMinusValue}.setValue(${nestedRightVal})`
    );
    return;
  }

  const leftVal = node.left?.name || node.argument?.name;
  let rightVal =
    node.right?.raw || src.slice(node.right?.start || 0, node.right?.end || 0);
  if (node.right?.value && node.right.value[0] === "$")
    rightVal = `${node.right.value}.getValue()`;

  const operation =
    node.operator === "++" ? "+ 1" : node.operator === "--" ? "- 1" : "";

  if (
    node.argument &&
    node.type === "UpdateExpression" &&
    node.argument.name[0] === "$"
  )
    rightVal = `${node.argument.name}.getValue() ${operation}`;

  if (
    node.right?.type === "UnaryExpression" &&
    node.right.argument.name[0] === "$"
  )
    rightVal = `${node.right.operator}${node.right.argument.name}.getValue()`;

  magicString.overwrite(
    node.start,
    node.end,
    `${leftVal}.setValue(${rightVal})`
  );
}

const arrProps = ["push", "pop", "splice", "shift", "unshift"];

function compileRezact(ast) {
  walk.ancestor(ast, {
    VariableDeclarator(node: any, _state) {
      const name = node.id.name;
      if (!name) return;
      if (name[0] === "$") {
        if (node.init.type === "Literal") wrapInUseSignal(node.init);
        if (node.init.type === "UnaryExpression") wrapInUseSignal(node.init);
        if (node.init.type === "ArrayExpression") wrapInUseMapState(node.init);
        if (node.init.type === "CallExpression")
          wrapInCreateComputed(node.init);
      }
    },

    TemplateLiteral(node: any, _state, ancestors) {
      if (isAttributeArg(ancestors)) {
        wrapInCreateComputedAttribute(node);
      } else {
        wrapInCreateComputed(node);
      }
    },

    UnaryExpression(node: any) {
      if (!node?.argument?.name) return;
      if (node.argument.name[0] === "$") {
      }
    },

    LogicalExpression(node: any, _state, ancestors: any) {
      if (node.left.type === "Identifier" && node.left.name[0] === "$") {
        tackOnDotVee(node.left);
      }

      const lastAncestor = ancestors.at(-1);
      const logicalExprAncestor = ancestors.find(
        (anc) =>
          anc.type === "LogicalExpression" ||
          anc.type === "ConditionalExpression"
      );

      if (
        logicalExprAncestor !== undefined &&
        logicalExprAncestor !== lastAncestor
      )
        return;

      const explicitDeps = findDependencies(node.left);
      wrapInCreateComputed(node, explicitDeps);
    },

    MemberExpression(node: any, _state, ancestors: any) {
      const propertyExcludeList = ["toJson", "map", "deleteValue", ...arrProps];
      let isInAssignment = false;
      let anceLen = ancestors.length - 1;
      while (anceLen > -1) {
        if (ancestors[anceLen].type === "MemberExpression") {
          anceLen--;
          continue;
        }
        if (ancestors[anceLen].type === "AssignmentExpression") {
          isInAssignment = true;
          break;
        }
        break;
      }

      if (!node.property || !node.property.type || !node.property.name) return;

      if (
        node.object.type === "Identifier" &&
        node.object.name[0] === "$" &&
        node.property.type === "Identifier" &&
        node.property.name[0] === "$"
      ) {
        tackOnDotVee(node.object);
      } else if (
        node.property.type === "Identifier" &&
        node.property.name[0] === "$"
      ) {
        if (
          ancestors.length > 2 &&
          ancestors.at(-2).callee?.property?.name === "xCreateElement"
        )
          return;
        if (
          ancestors.length > 2 &&
          ancestors.at(-2).callee?.name === "xCreateElement"
        )
          return;
        if (isInAssignment)
          return wrapInSetValue(ancestors[anceLen], node.property);
        tackOnDotVee(node.property);
      } else if (
        node.object.name &&
        node.object.name[0] === "$" &&
        !propertyExcludeList.includes(node.property.name)
      ) {
        tackOnDotVee(node.object);
        if (node.computed) tackOnDotVee(node);
      }
    },

    CallExpression(node: any) {
      if (node.arguments) {
        node.arguments.forEach((arg) => {
          if (
            arg.type === "Identifier" &&
            arg.name[0] === "$" &&
            node.callee?.object?.name === "console" &&
            node.callee?.property?.name === "log"
          ) {
            tackOnDotVee(arg);
          } else if (
            node.callee.type === "MemberExpression" &&
            arrProps.includes(node.callee.property.name) &&
            arg.type === "Identifier" &&
            arg.name[0] === "$"
          ) {
            tackOnDotVee(arg);
          }
        });
      }
      if (
        node.callee.type === "MemberExpression" &&
        node.callee.object.type === "Identifier" &&
        node.callee.object.name[0] === "$" &&
        node.callee.property.name === "map"
      ) {
        magicString.overwrite(
          node.callee.property.start,
          node.callee.property.end,
          "Map"
        );
        wrapInCreateMapped(node, [node.callee.object.name]);
      }
    },

    Identifier(_node: any, _state) {
      if (_node.name === "onMount") {
        importsUsed.useCustomElementsForMountCallbacks = true;
        functionsToRun.push("useCustomElementsForMountCallbacks()");
      }
      if (_node.name === "onUnmount") {
        importsUsed.useCustomElementsForMountCallbacks = true;
        functionsToRun.push("useCustomElementsForMountCallbacks()");
      }
    },

    Property(node: any, _state) {
      if (node.key.name === "className") {
        magicString.overwrite(node.key.start, node.key.end, "class");
      }
      if (node.key.name && node.key.name[0] === "$") {
        wrapInUseSignal(node.value);
      }
      if (node.key.value && node.key.value[0] === "$") {
        wrapInUseSignal(node.value);
      }
      if (node.key.name === "value" && !importsUsed.useInputs)
        importsUsed.useInputs = true && functionsToRun.push("useInputs()");
    },

    ConditionalExpression(node: any, _state, ancestors: any) {
      if (
        isAttributeArg(ancestors) &&
        !hasAncestor(ancestors, "TemplateLiteral")
      ) {
        wrapInCreateComputedAttribute(node);
      } else if (!hasAncestor(ancestors, "TemplateLiteral")) {
        wrapInCreateComputed(node);
      }
      if (!node?.test?.name) return;
      if (node.test.name && node.test.name[0] === "$") {
        tackOnDotVee(node.test);
      }
    },

    ImportDeclaration(node: any) {
      lastImport = node;
    },

    BinaryExpression(node: any) {
      const name = node.left.name;
      if (name && name[0] === "$") {
        tackOnDotVee(node.left);
      }
    },

    UpdateExpression(node: any, _state) {
      const name = node.argument.name;
      if (!name) return;
      if (name[0] === "$") {
        wrapInSetValue(node);
      }
    },

    AssignmentExpression(node: any, _state) {
      const name = node.left.name;
      if (!name) return;
      if (name[0] === "$") {
        wrapInSetValue(node);
      }
    },

    LabeledStatement(node: any) {
      if (node.label.name === "$") {
        if (node.body.type === "BlockStatement")
          wrapInCreateComputed(node.body);
        if (node.body.type === "ExpressionStatement")
          wrapInCreateComputed(node.body.expression);
      }
    },
  });
}

function rezact(): PluginOption {
  return {
    name: "transform-rezact",
    // enforce: "pre",
    transform(_src, id) {
      const supportedFileTypes = [".tsx", ".jsx", ".ts", ".js", ".mdx"];
      if (supportedFileTypes.find((sf) => id.includes(sf)) === undefined)
        return;
      if (id.includes("node_modules")) return;
      if (id.includes("rezact.ts")) return;
      if (id.includes("signals.ts")) return;
      if (id.includes("mapState.ts")) return;
      src = _src;
      magicString = null;
      lastImport = null;
      importsUsed = {};
      functionsToRun = [];
      // console.log(id);
      magicString = new MagicString(src);
      const ast = acorn.parse(src, {
        locations: true,
        ecmaVersion: "latest",
        sourceType: "module",
      });
      compileRezact(ast);
      const importsUsedArr = Object.keys(importsUsed);
      if (importsUsedArr.length > 0)
        magicString.prepend(
          `import {${importsUsedArr.join(",")}} from "src/lib/rezact/rezact"\n`
        );

      const signalsUsedArr = Object.keys(signalsUsed);
      if (signalsUsedArr.length > 0)
        magicString.prepend(
          `import {${signalsUsedArr.join(",")}} from "src/lib/rezact/signals"\n`
        );

      const mapStateUsedArr = Object.keys(mapStateUsed);
      if (mapStateUsedArr.length > 0)
        magicString.prepend(
          `import {${mapStateUsedArr.join(
            ","
          )}} from "src/lib/rezact/mapState"\n`
        );

      if (lastImport)
        magicString.appendRight(
          lastImport.end,
          `\n${functionsToRun.join("\n")}`
        );
      return {
        code: magicString.toString(),
        map: magicString.generateMap({ hires: true }),
      };
    },
  };
}

export { rezact };
