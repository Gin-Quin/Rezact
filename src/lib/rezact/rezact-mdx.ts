import { xCreateElement, xFragment } from "./rezact";

export default { xCreateElement, xFragment };
