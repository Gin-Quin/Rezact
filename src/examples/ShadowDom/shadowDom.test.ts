import { describe, expect, it } from "vitest";
import { render } from "rezact";
import { Page } from "./ShadowDomForTesting";

describe("Shadow DOM suite", () => {
  it("Renders an Element with scoped CSS using shadow DOM", () => {
    render(document.body, Page);
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it("renders a div element with a shadow root", () => {
    const div = document.querySelector("div");
    expect(div.shadowRoot.innerHTML).toMatchSnapshot();
  });
});
