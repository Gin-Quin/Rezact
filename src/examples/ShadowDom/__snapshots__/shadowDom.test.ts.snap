// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shadow DOM suite > Renders an Element with scoped CSS using shadow DOM 1`] = `"<div></div><h1>asdfasdf</h1>"`;

exports[`Shadow DOM suite > renders a div element with a shadow root 1`] = `"<style></style><h1>Shadow DOM Test (Scoped CSS)</h1><p>asdfasdf</p>"`;
